Update for Julia 0.6

samoconnor opened the pull request.

roberthoenig commented:
Excited for this! :)

samoconnor commented:
As it stands, everything in `runtests.jl` is passing on my dev machine.
I need to do a bunch of doc updates.
I also plan to add a working example of using a local docker image to prepare a Lambda with arbitrary binary dependancies.

If you're feeling brave you can try pulling down this branch and then doing:

```julia
julia> using AWSLambda

julia> AWSLambda.delete_lambda("jl_lambda_eval")

julia> AWSLambda.deploy_jl_lambda_base()
Dict{Symbol,Any} with 6 entries:
  :RoutingConfig   => nothing
  :Description     => ""
  :AliasArn        => "arn:aws:lambda:ap-southeast-2:551613799374:function:jl_lambda_eval:JL_0_6_2"
  :FunctionVersion => "17"
  :RevisionId      => "92d61df1-01af-4ee2-ab0d-9b04d8732945"
  :Name            => "JL_0_6_2"

julia> AWSCore.set_debug_level(1)

julia> @lambda_eval readstring(`uname -a`)
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

"Linux ip-10-11-243-52 4.9.81-35.56.amzn1.x86_64 #1 SMP Fri Feb 16 00:18:48 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux\n"

julia> Pkg.test("AWSLambda")
```

Reading `test/runtests.jl` will give you an idea of the API.

roberthoenig commented:
@samoconnor I get an error when trying to import the package:
```julia
julia> Pkg.checkout("AWSLambda", "06_branch")
INFO: Checking out AWSLambda 06_branch...
INFO: Pulling AWSLambda latest 06_branch...
INFO: No packages to install, update or remove

julia> using AWSLambda
INFO: Recompiling stale cache file /home/robert/.julia/lib/v0.6/AWSLambda.ji for module AWSLambda.
ERROR: LoadError: 404 -- Function not found: arn:aws:lambda:us-east-1:959667734437:function:jl_lambda_eval:JL_0_6_2
HTTP.ExceptionRequest.StatusError(404, HTTP.Messages.Response:
"""
HTTP/1.1 404 Not Found
Date: Tue, 27 Mar 2018 15:24:01 GMT
Content-Type: application/json
Content-Length: 118
Connection: keep-alive
x-amzn-RequestId: e0bbb04f-31d2-11e8-b36d-15e4682238a1
x-amzn-ErrorType: ResourceNotFoundException

{"Message":"Function not found: arn:aws:lambda:us-east-1:959667734437:function:jl_lambda_eval:JL_0_6_2","Type":"User"}""")

Stacktrace:
 [1] include_from_node1(::String) at ./loading.jl:576
 [2] include(::String) at ./sysimg.jl:14
 [3] anonymous at ./<missing>:2
while loading /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl, in expression starting on line 812
ERROR: Failed to precompile AWSLambda to /home/robert/.julia/lib/v0.6/AWSLambda.ji.
Stacktrace:
 [1] compilecache(::String) at ./loading.jl:710
 [2] _require(::Symbol) at ./loading.jl:463
 [3] require(::Symbol) at ./loading.jl:405

julia> 
```

samoconnor commented:
There is a chicken/egg problem. A marco is trying to execute a lambda before it is installed.
I will get back to this properly in a couple of days.
In the meantime, the workaround is to comment-out this line:
https://github.com/samoconnor/AWSLambda.jl/blob/786af832ece994c103f0ca24b9bcc8f535fcfb71/src/AWSLambda.jl#L684-L685

roberthoenig commented:
Yay, things seem to be working: I successfully uploaded the lambda and ran the uname command on it. The tests are failing, however:
```
robert@robhoenig:~/.julia$ julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.2 (2017-12-13 18:08 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-pc-linux-gnu

julia> using AWSLambda
INFO: Precompiling module AWSLambda.

julia> AWSLambda.deploy_jl_lambda_base()
Dict{Symbol,Any} with 6 entries:
  :RoutingConfig   => nothing
  :Description     => ""
  :AliasArn        => "arn:aws:lambda:us-east-1:959667734437:function:jl_lambda_eval:JL_0_6_2"
  :FunctionVersion => "1"
  :RevisionId      => "72c54e98-9b48-41b6-91cb-bc8d9d9e5ce5"
  :Name            => "JL_0_6_2"

julia> AWSCore.set_debug_level(1)
1

julia> @lambda_eval readstring(`uname -a`)
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

"Linux ip-10-20-51-48 4.9.85-38.58.amzn1.x86_64 #1 SMP Wed Mar 14 01:17:26 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux\n"

julia> Pkg.test("AWSLambda")
INFO: Testing AWSLambda
Loading "default" AWSCredentials from /home/robert/.aws/credentials... (AKIAJW6HRNYSPORXDZVQ, Jks...)

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
21 primes between 10 and 100.

lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
348509 primes between 10 and 5000010.

316066 primes between 5000010 and 10000010.

299904 primes between 15000010 and 20000010.

291932 primes between 25000010 and 30000010.

306125 primes between 10000010 and 15000010.

282974 primes between 45000010 and 50000010.

278439 primes between 60000010 and 65000010.

295320 primes between 20000010 and 25000010.

288915 primes between 30000010 and 35000010.

275077 primes between 75000010 and 80000010.

286880 primes between 35000010 and 40000010.

284505 primes between 40000010 and 45000010.

272795 primes between 90000010 and 95000010.

281066 primes between 50000010 and 55000010.

276241 primes between 70000010 and 75000010.

279915 primes between 55000010 and 60000010.

277510 primes between 65000010 and 70000010.

273221 primes between 85000010 and 90000010.

274351 primes between 80000010 and 85000010.

271706 primes between 95000010 and 100000000.

5761451 primes between 10 and 100000000.
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
Error During Test
  Test threw an exception of type AWSLambda.AWSLambdaException
  Expression: count_primes2(10, 100) == 21
  AWSLambdaException "jl_lambda_eval:JL_0_6_2":
  EOF on Julia stdout! Exit code: 1
  ERROR: UndefVarError: Primes not defined
  Stacktrace:
   [1] #136 at /home/robert/.julia/v0.6/AWSLambda/test/runtests.jl:148 [inlined]
   [2] (::Base.Serializer.__deserialized_types__.##138#139)() at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:617
   [3] lambda_function(::Function) at /var/task/julia/v0.6/module_jl_lambda_eval.jl:18
   [4] invoke_jl_lambda(::Module, ::String, ::IOStream) at /var/task/julia/v0.6/AWSLambdaWrapper.jl:25
   [5] invoke_lambda(::Module, ::Dict{String,Any}, ::IOStream) at /var/task/julia/v0.6/AWSLambdaWrapper.jl:33
   [6] open(::AWSLambdaWrapper.##1#2{Module}, ::String, ::String) at ./iostream.jl:152
   [7] main(::Module) at /var/task/julia/v0.6/AWSLambdaWrapper.jl:50
   [8] process_options(::Base.JLOptions) at ./client.jl:286
   [9] _start() at ./client.jl:371
  Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz
```

samoconnor commented:
Do you have the Primes package installed locally?
Have you uncommented line 684?

roberthoenig commented:
I did uncomment line 684 and Primes is installed locally. From looking at the exception traceback, it seems like the error was thrown remotely in the lambda? Is it possible that Primes needs to be added there?

samoconnor commented:
The failed test is intended to: bundle up the source files from the local Primes package; send them to Lambda; write them out to files under /tmp; and make them available to the Julia runtime on Lambda. (That is what “embed module files in body” on line 684 does.)
I’ve made some local changes to the way that works to make it precompile compatible, it might be best for you to wait for me to commit/push.
Thanks for taking the time to try it out.

roberthoenig commented:
Another thing is that this doesn't work:
```shell
julia> aws = aws_config()
Loading "default" AWSCredentials from /home/robert/.aws/credentials... (AKIAJW6HRNYSPORXDZVQ, Jks...)


Dict{Symbol,Any} with 2 entries:
  :creds  => (myapikey, Jks...)…
  :region => "us-east-1"

julia> λ = @lambda aws function sum(a::Int, b::Int)
           return a+b
       end
ERROR: ArgumentError: function requires all((x->begin  # /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl, line 657:
            x isa Expr
        end), args[1:end - 1])

julia> 
```

samoconnor commented:
Yes, the argument order has changed. The current docstring specifies that the "aws" arg should come at the end:
```julia
help?> AWSLambda.@lambda
  @lambda [using ...] function name(args...) body end [aws_config]

  Define local function that executes body on AWS Lambda.
```

Also note that you can omit the aws config argument and the config from your credentials file will be used by default. The `runtests.jl` file has examples of this.

If you prefer, you can pass the aws config as the last argument to the `@lambda` macro like this:

```julia
julia> AWSCore.set_debug_level(2)
2

julia> aws = AWSCore.aws_config()
Loading "octech" AWSCredentials from /Users/sam/.aws/credentials... (XXXX, XXX...)


Dict{Symbol,Any} with 2 entries:
  :creds  => (XXXX, XXX...)…
  :region => "ap-southeast-2"

julia> @lambda function mysum(a::Int, b::Int)
           return a+b
       end aws
mysum (generic function with 1 method)

julia> mysum(1, 2)
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations HTTP/1.1
HTTP/1.1 200 OK <= (POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations HTTP/1.1)
HTTP.Messages.Response:
"""
HTTP/1.1 200 OK
Date: Tue, 03 Apr 2018 03:36:34 GMT
Content-Type: application/json
Content-Length: 33
Connection: keep-alive
x-amzn-RequestId: 35481fac-36f0-11e8-9b02-9b86dfb0fce4
x-amzn-Remapped-Content-Length: 0
X-Amz-Executed-Version: 17
X-Amzn-Trace-Id: root=1-5ac2f6c2-e157502c6ad8ba7c39a90ebc;sampled=0

{"jl_data": "yA==", "stdout": ""}"""


3

julia>
```

roberthoenig commented:
Thanks for helping me out with this!

roberthoenig commented:
I hope I'm not pestering you too much with troubleshooting; I encountered another issue I cannot resolve:
```shell
julia> @deploy_lambda function mysum(a::Int, b::Int)
       return a+b
       end aws
full_load_path = ""
py_config = String["os.environ['JULIA_LOAD_PATH'] += ''\n"]
create_jl_lambda(mysum)
py_config = String["os.environ['JULIA_LOAD_PATH'] += ''\n", "error_sns_arn = ''\n"]
query Dict{Any,Any}(Pair{Any,Any}(:headers, Dict{Any,Any}()))
Dict{Symbol,Any}(Pair{Symbol,Any}(:verb, "GET"),Pair{Symbol,Any}(:version, "2015-03-31"),Pair{Symbol,Any}(:args, Dict{Any,Any}(Pair{Any,Any}(:headers, Dict{Any,Any}()))),Pair{Symbol,Any}(:region, "us-east-1"),Pair{Symbol,Any}(:url, "https://lambda.us-east-1.amazonaws.com/2015-03-31/functions/mysum/configuration"),Pair{Symbol,Any}(:service, "lambda"),Pair{Symbol,Any}(:ordered_json_dict, false),Pair{Symbol,Any}(:headers, Dict("Host"=>"lambda.us-east-1.amazonaws.com","Content-Type"=>"application/json","User-Agent"=>"AWSCore.jl/0.0.0")),Pair{Symbol,Any}(:creds, (AKIAJW6HRNYSPORXDZVQ, Jks...)
),Pair{Symbol,Any}(:content, "{\"headers\":{}}"),Pair{Symbol,Any}(:resource, "/2015-03-31/functions/mysum/configuration"))
lambda.GET /2015-03-31/functions/mysum/configuration
GET /2015-03-31/functions/mysum/configuration HTTP/1.1
GET /2015-03-31/functions/mysum/configuration HTTP/1.1
HTTP/1.1 404 Not Found <= (GET /2015-03-31/functions/mysum/configuration HTTP/1.1)
query Dict{Symbol,Any}(Pair{Symbol,Any}(:jl_data, "very long base64 string here"}"),Pair{Symbol,Any}(:resource, "/2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations"))
lambda.POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations
POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations HTTP/1.1
HTTP/1.1 200 OK <= (POST /2015-03-31/functions/jl_lambda_eval%3AJL_0_6_2/invocations HTTP/1.1)
HTTP.Messages.Response:
"""
HTTP/1.1 200 OK
Date: Tue, 03 Apr 2018 04:06:44 GMT
Content-Type: application/json
Content-Length: 1302
Connection: keep-alive
x-amzn-RequestId: 5d6a076a-36f4-11e8-bca6-a501ae0ff9fd
X-Amz-Function-Error: Unhandled
x-amzn-Remapped-Content-Length: 0
X-Amz-Executed-Version: 1
X-Amzn-Trace-Id: root=1-5ac2fdbc-d7fe233aa15eec92a750a3d2;sampled=0

{"stackTrace": [["/var/task/lambda_function.py", 101, "lambda_handler", "raise Exception(message + '\\n' + out + cpu_model_name)"]], "errorType": "Exception", "errorMessage": "EOF on Julia stdout! Exit code: 1\nFoo\npush!(LOAD_PATH, \"/tmp/tmpip0wM4\")\ninsert!(Base.LOAD_CACHE_PATH, 1, \"/tmp/tmpip0wM4/julia/lib/v0.6\")\nusing module_mysum\n\nERROR: UndefVarError: create_lambda not defined\nStacktrace:\n [1] (::Base.Serializer.__deserialized_types__.##23#25)(::Dict{Symbol,Any}, ::String, ::Array{Any,1}, ::Dict{Symbol,Any}, ::Bool) at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:864\n [2] (::Base.Serializer.__deserialized_types__.##27#28)() at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:617\n [3] lambda_function(::Function) at /var/task/julia/v0.6/module_jl_lambda_eval.jl:18\n [4] invoke_jl_lambda(::Module, ::String, ::IOStream) at /var/task/julia/v0.6/AWSLambdaWrapper.jl:25\n [5] invoke_lambda(::Module, ::Dict{String,Any}, ::IOStream) at /var/task/julia/v0.6/AWSLambdaWra
⋮
1302-byte body
"""
HTTP.Messages.Response:
"""
HTTP/1.1 200 OK
Date: Tue, 03 Apr 2018 04:06:44 GMT
Content-Type: application/json
Content-Length: 1302
Connection: keep-alive
x-amzn-RequestId: 5d6a076a-36f4-11e8-bca6-a501ae0ff9fd
X-Amz-Function-Error: Unhandled
x-amzn-Remapped-Content-Length: 0
X-Amz-Executed-Version: 1
X-Amzn-Trace-Id: root=1-5ac2fdbc-d7fe233aa15eec92a750a3d2;sampled=0

{"stackTrace": [["/var/task/lambda_function.py", 101, "lambda_handler", "raise Exception(message + '\\n' + out + cpu_model_name)"]], "errorType": "Exception", "errorMessage": "EOF on Julia stdout! Exit code: 1\nFoo\npush!(LOAD_PATH, \"/tmp/tmpip0wM4\")\ninsert!(Base.LOAD_CACHE_PATH, 1, \"/tmp/tmpip0wM4/julia/lib/v0.6\")\nusing module_mysum\n\nERROR: UndefVarError: create_lambda not defined\nStacktrace:\n [1] (::Base.Serializer.__deserialized_types__.##23#25)(::Dict{Symbol,Any}, ::String, ::Array{Any,1}, ::Dict{Symbol,Any}, ::Bool) at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:864\n [2] (::Base.Serializer.__deserialized_types__.##27#28)() at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:617\n [3] lambda_function(::Function) at /var/task/julia/v0.6/module_jl_lambda_eval.jl:18\n [4] invoke_jl_lambda(::Module, ::String, ::IOStream) at /var/task/julia/v0.6/AWSLambdaWrapper.jl:25\n [5] invoke_lambda(::Module, ::Dict{String,Any}, ::IOStream) at /var/task/julia/v0.6/AWSLambdaWra
⋮
1302-byte body
"""

ERROR: AWSLambdaException "jl_lambda_eval:JL_0_6_2":
EOF on Julia stdout! Exit code: 1
Foo
push!(LOAD_PATH, "/tmp/tmpip0wM4")
insert!(Base.LOAD_CACHE_PATH, 1, "/tmp/tmpip0wM4/julia/lib/v0.6")
using module_mysum

ERROR: UndefVarError: create_lambda not defined
Stacktrace:
 [1] (::Base.Serializer.__deserialized_types__.##23#25)(::Dict{Symbol,Any}, ::String, ::Array{Any,1}, ::Dict{Symbol,Any}, ::Bool) at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:864
 [2] (::Base.Serializer.__deserialized_types__.##27#28)() at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:617
 [3] lambda_function(::Function) at /var/task/julia/v0.6/module_jl_lambda_eval.jl:18
 [4] invoke_jl_lambda(::Module, ::String, ::IOStream) at /var/task/julia/v0.6/AWSLambdaWrapper.jl:25
 [5] invoke_lambda(::Module, ::Dict{String,Any}, ::IOStream) at /var/task/julia/v0.6/AWSLambdaWrapper.jl:33
 [6] open(::AWSLambdaWrapper.##1#2{Module}, ::String, ::String) at ./iostream.jl:152
 [7] main(::Module) at /var/task/julia/v0.6/AWSLambdaWrapper.jl:50
 [8] process_options(::Base.JLOptions) at ./client.jl:286
 [9] _start() at ./client.jl:371
Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz


Stacktrace:
 [1] #invoke_lambda#10(::Bool, ::Function, ::Dict{Symbol,Any}, ::String, ::Dict{Symbol,Any}) at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:317
 [2] (::AWSLambda.#kw##invoke_lambda)(::Array{Any,1}, ::AWSLambda.#invoke_lambda, ::Dict{Symbol,Any}, ::String) at ./<missing>:0
 [3] invoke_jl_lambda(::Dict{Symbol,Any}, ::String, ::Function, ::Vararg{Function,N} where N) at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:596
 [4] (::AWSLambda.##34#35{Dict{Symbol,Any},##23#25})(::Dict{Symbol,Any}, ::Vararg{Any,N} where N) at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:616
 [5] deploy_lambda(::Dict{Symbol,Any}, ::String, ::Array{Any,1}, ::Dict{Symbol,Any}, ::Bool) at ./<missing>:0
 [6] create_jl_lambda(::Dict{Symbol,Any}, ::String, ::String, ::Array{Symbol,1}, ::Dict{Symbol,Any}) at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:805
 [7] create_jl_lambda(::Dict{Symbol,Any}, ::String, ::String, ::Array{Symbol,1}) at /home/robert/.julia/v0.6/AWSLambda/src/AWSLambda.jl:737
```

samoconnor committed:
Add default_lambda_module_cache so that precompiled_module_files
works before deploy_jl_lambda_base.

samoconnor commented:
@roberthoenig the `UndefVarError: create_lambda not defined` error should be resolved in the latest commit.

I recommend doing `AWSLambda.delete_lambda("jl_lambda_eval")` and then `AWSLambda.deploy_jl_lambda_base()` to ensure that everything is up to date.

samoconnor committed:
moved base Dockerfile to https://github.com/samoconnor/lambdajl

samoconnor committed:
move make.jl to lambdajlbuild/make.jl

samoconnor committed:
dont try to build create_jl_lambda into sysimg

samoconnor committed:
dockerfile reorg

samoconnor committed:
example of custom REQUIRE file

samoconnor committed:
README updates (incomplete)

samoconnor committed:
revise jl_lambda_custom/make.jl

roberthoenig commented:
You mean `AWSLambda.deploy_jl_lambda_eval()`?

samoconnor commented:
Right, it changed since I wrote that :)

roberthoenig commented:
Everything is working for me now, thanks!

However, I wasn't able to use Amazons API gateway with a lambda deployed by AWSLambda.jl. The reason for that is that the JSON returned by `invoke_lambda` does not comply with the API gateway format:
> Resolution
For a Lambda function's response to be handled the by API Gateway, it must return a response in this format:
{
    "isBase64Encoded": true|false,
    "statusCode": httpStatusCode,
    "headers": { "headerName": "headerValue", ... },
    "body": "..."
}

Afaik, `invoke_lambda` returns a JSON that looks like the following:
```
{"jl_data": "", "stdout": ""}
```
Is there anything I'm misunderstanding here?

samoconnor commented:
> Everything is working for me now, thanks!

Good to know!

Re API gateway, I had a toy example of that working long ago but I haven't tried it recently.

This code might help you to see what input/ouput configuration I used. See the part about `responseModels` and `responseTemplates`.
https://github.com/samoconnor/AWSLambda.jl/blob/c430540fd95a0a6b187fb3d85e5dc7537e3758f3/src/AWSLambda.jl#L916-L944 

FYI, This is the place where the output is generated:
https://github.com/samoconnor/AWSLambda.jl/blob/e1ec18179ef5684ffebe9c5341227fa9a28a4e62/docker/jl_lambda_eval/lambda_function.py#L139-L144

It should be simple enough to provide an option to return the JSON produced by the Julia function directly (without the "jl_data" wrapper).

samoconnor commented:
... in fact, I think that since I'm making API changes in this release, It might be best to return un-wrapped JSON by default. I can still return the `{"jl_data": "", "stdout": ""}` format when the request comes from julia in the form of base64 encoded serialised Julia objects, but if the request comes as JSON i'll just return un-wrapped JSON.

samoconnor commented:
Also, FYI, I'm working on an example of building a custom jl_lambda_eval image with custom built-in packages.
See https://github.com/samoconnor/AWSLambda.jl/tree/06_branch/docker/jl_lambda_custom
The REQUIRE file specifies the packages needed for your project.
The following commands should result in the base `jl_lambda_eval` function being replaced by a custom version.
```bash
$ julia make.jl build
$ julia make.jl zip
$ julia make.jl deploy
```
The REQUIRE file in the example includes `JuMP` and `Clp`.
After this is deployed I can do:
```julia
julia> using AWSLambda, JuMP, Clp

julia> @lambda using Clp, JuMP function foo()
           m = Model(solver = ClpSolver())
           @variable(m, 0 <= x <= 2)
           @variable(m, 0 <= y <= 30)
           @objective(m, Max, 5x + 3y)
           @constraint(m, 1x + 5y <= 3.0)
           print(m)
           status = solve(m)
           println("Objective value: ", getobjectivevalue(m))
           println("x = ", getvalue(x))
           println("y = ", getvalue(y))
           return m
       end
foo (generic function with 1 method)

julia> foo()
Max 5 x + 3 y
Subject to
 x + 5 y ≤ 3
 0 ≤ x ≤ 2
 0 ≤ y ≤ 30
Objective value: 10.6
x = 2.0
y = 0.2

Maximization problem with:
 * 1 linear constraint
 * 2 variables
Solver is ClpMathProg
```

samoconnor committed:
remove empty mod_files shortcut from embed_modules_in_body to ensure that using is still executed

samoconnor committed:
rename deploy_jl_lambda_base -> deploy_jl_lambda_eval

roberthoenig commented:
Sounds cool! I got my workaround by using a dedicated aws user with narrowed-down permissions instead of the api for now.

samoconnor commented:
In all my usage of AWS Lambda I've never ended up using API Gateway.
I've always just used temporary AWS credentials with fine-grained permissions to call the lambdas directly from the client.
I've built a few simple serverless web apps where:
 - static HTML/JS is served from S3
 - the user authenticates with Cognito User Pools via the AWS JS SDK
 - the browser js code then uses the AWS JS SDK to call lambdas directly using the credentials obtained from cognito

I can see that API Gateway might be useful if you need to make an API publicly available for 3rd party users, but I haven't had that need yet.

samoconnor committed:
- Remove obsolete error SNS stuff. Better to use the new `DeadLetterConfig` feature.

 - Return raw result JSON instead of wrapping with `jl_data` and
   `stdout`.
   (stdout now only returned for Julia sieraliased object invocations).

 - Add extra metadata to exception messages to simplify finding logs
   in cloudwatch:
    invoked_function_arn
    function_version
    aws_request_id
    log_group_name
    log_stream_name
    cpu

 - Update README.md

samoconnor committed:
Improve error message
